### PR TITLE
BinaryTree Traits recorridos inorder postorder preorder iterators macro

### PR DESCRIPTION
**Pull request by palominor**
Estimado dr. Cuadros
Adjunto los cambios realizados a binaryTree, los recorridos con iteradores, el cambio a traits y el uso de una macro
image
Ejecucion cambios de generalizar el print (forward iterator) (print ahora solo usa el forward_iterator, es una utilidad y print sigue funcionando pero solo aplicando el general iterator, print es un utility/wrapper alrededor del iterator generalizado)
image

### CODE REVIEW

File: binarytree.h [hunk 1]
```diff
@@ -1,168 +1,418 @@
...
+};
 
-    bool getPrinted(){return m_printed;}
-    void setPrinted(bool printed){m_printed = printed;}
-    size_t getLevel(){return m_level;}
+//*****************************************************************************
+//*                               Iterators                                   *
+//*****************************************************************************
+template <typename Container>
+class binary_tree_iterator : public general_iterator<Container, class binary_tree_iterator<Container>> //
```
Comment (edson2495) [inline binarytree.h:54]: You might wanna remove the inheritance.

File: binarytree.h [hunk 2]
```diff
@@ -1,168 +1,418 @@
...
-    // TODO: generalize this function by using iterators and apply any function
-    void print(ostream &os, void (*func) (Node& node, ostream &os)){
-        foreach(printbegin(),printend(), func, os);
-        foreach(inbegin(1),inend(1), [](Node& node){});
+    // DONE: generalize this function by using iterators and apply any function
+    // Se mantiene esta pero se crea un iterador que recorre el arbol
+    // En el orden en que se imprime. Lo que un iterador conoce es
+    // El inicio, final y el valor. No conoce el nivel. Por lo que esta funcion
+    // Aun es util
+    void print(Node  *pNode, ostream &os, size_t level)
```
Comment (edson2495) [inline binarytree.h:371]: That's not true. You CAN generalize this function, you might wanna try in order to practice.

**Review by edson2495 — APPROVED** [review]
Partialy approved.

File: binarytree.h (hunk 1 above)
Comment (palominor) [reply binarytree.h:54]: Thank you for your feedback, solved in https://github.com/ecuadros/EDA/pull/103, please check it out. Inheritance was removed

File: binarytree.h (hunk 2 above)
Comment (palominor) [reply binarytree.h:371]: Thank you for your feedback, solved in https://github.com/ecuadros/EDA/pull/103, please check it out. Generalized the print function using an iterator

**Pull request closed without merging**